controllers/metal3.io/preprovisioningimage_controller.go

```diff
@@ -90,56 +117,104 @@ func (r *PreprovisioningImageReconciler) Reconcile(ctx context.Context, req ctrl
+	networkData, secretStatus, err := getNetworkData(secretManager, img)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			log.Info("network data Secret does not exist")
+			return setError(generation, &img.Status, reasonImageMissingNetworkData, "NetworkData secret not found"), err
+		}
+		return false, err
 	}
 
-	if k8serrors.IsNotFound(err) {
-		log.Info("network data Secret does not exist")
-		return setError(generation, &img.Status, reasonImageMissingNetworkData, "NetworkData secret not found"), err
+	if configChanged(img, format, secretStatus) {
+		reason := "Config changed"
+		if meta.IsStatusConditionTrue(img.Status.Conditions, string(metal3.ConditionImageReady)) {
+			// Ensure we mark the status as not ready before we remove the build
+			// from the image cache.
+			setUnready(generation, &img.Status, reason)
+		} else {
+			if img.Status.Format != "" {
+				if err := r.discardExistingImage(img, log); err != nil {
+					return false, err
+				}
+			}
+			// Set up all the data before building the image and adding the URL,
+			// so that even if we fail to write the built image status and the
+			// config subsequently changes, the image cache cannot leak.
+			setImage(generation, &img.Status, "", format,
+				secretStatus, img.Spec.Architecture,
+				reason)
+		}
+		return true, nil
 	}
 
-	return false, err
+	var networkDataContent imageprovider.NetworkData
+	if networkData != nil {
+		networkDataContent = networkData.Data
+	}
+	url, err := r.ImageProvider.BuildImage(imageprovider.ImageData{
+		ImageMetadata:     img.ObjectMeta.DeepCopy(),
+		Format:            format,
+		Architecture:      img.Spec.Architecture,
+		NetworkDataStatus: secretStatus,
+	}, networkDataContent, log)
+	if err != nil {
+		return false, err
+	}
+	log.Info("image URL available", "url", url, "format", format)
+
+	return setImage(generation, &img.Status, url, format,
+		secretStatus, img.Spec.Architecture,
+		"Generated image"), nil
 }
 
-func getImageURL(acceptFormats []metal3.ImageFormat) (url string, format metal3.ImageFormat, errorMessage string) {
-	for _, fmt := range acceptFormats {
-		switch fmt {
-		case metal3.ImageFormatISO:
-			if iso := os.Getenv("DEPLOY_ISO_URL"); iso != "" {
-				return iso, fmt, ""
-			}
-			if errorMessage == "" {
-				format = fmt
-				errorMessage = "No DEPLOY_ISO_URL specified"
-			}
-		case metal3.ImageFormatInitRD:
-			if initrd := os.Getenv("DEPLOY_RAMDISK_URL"); initrd != "" {
-				return initrd, fmt, ""
-			}
-			if errorMessage == "" {
-				format = fmt
-				errorMessage = "No DEPLOY_RAMDISK_URL specified"
-			}
+func (r *PreprovisioningImageReconciler) getImageFormat(spec metal3.PreprovisioningImageSpec, log logr.Logger) (format metal3.ImageFormat) {
+	for _, acceptableFormat := range spec.AcceptFormats {
+		if r.ImageProvider.SupportsFormat(acceptableFormat) {
+			return acceptableFormat
 		}
 	}
+
+	if len(spec.AcceptFormats) > 0 {
+		log = log.WithValues("preferredFormat", spec.AcceptFormats[0])
```

Comment:
Should return after this?. Otherwise, it will log `no acceptable image format supported` in any case.

Reply:
That's the idea. If there were any acceptable formats supported, we'd have returned on line 194.

Note that this line doesn't actually log anything, it's just adding more data to the log message below. I'd have done it inline but it has to be conditional because getting `spec.AcceptFormats[0]` would fail if the slice is empty.